hikari/traits.py

```diff
@@ -446,4 +446,177 @@ async def update_presence(
 class BotAware(RESTAware, ShardAware, EventFactoryAware, DispatcherAware, typing.Protocol):
     """Structural supertype for a component that is aware of all internals."""
 
+    async def join(self, until_close: bool = True) -> None:
+        """Wait indefinitely until the application closes.
+
+        This can be placed in a task and cancelled without affecting the
+        application runtime itself. Any exceptions raised by shards will be
+        propagated to here.
+
+        Other Parameters
+        ----------------
+        until_close : builtins.bool
+            Defaults to `builtins.True`. If set, the waiter will stop as soon as
+            a request for shut down is processed. This can allow you to break
+            and begin closing your own resources.
+
+            If `builtins.False`, then this will wait until all shards' tasks
+            have died.
+        """
+        raise NotImplementedError
+
+    def run(
+        self,
+        *,
+        activity: typing.Optional[presences.Activity] = None,
+        afk: bool = False,
+        asyncio_debug: typing.Optional[bool] = None,
+        close_executor: bool = False,
+        close_loop: bool = True,
+        coroutine_tracking_depth: typing.Optional[int] = None,
+        enable_signal_handlers: bool = True,
+        idle_since: typing.Optional[datetime.datetime] = None,
+        ignore_session_start_limit: bool = False,
+        large_threshold: int = 250,
+        status: presences.Status = presences.Status.ONLINE,
+        shard_ids: typing.Optional[typing.Set[int]] = None,
+        shard_count: typing.Optional[int] = None,
```

Comment:
I don't agree that most of these should be present.

Half of this stuff is implementation specific, and implementations may not want to or may not be able to provide some of these.

I'd revert:
- asyncio_debug  (as their loop may not support this properly or may have other implications if it uses an existing loop)
- close_loop  (as their impl may use the event loop outside this)
- coroutine_tracking_depth (provisional feature in CPython, may go at any time)
- enable_signal_handlers (implementations may not bother with signals at all)

Reply:
ok 👍

Reply:
Under the rule of `def x(a, b, c)` being compatible with `def x(a, b, c, d=None, e=None, f=None), this should be fine with MyPy.

Reply:
Also afaik unfortunately you'll have to duplicate this docstring because pdoc3 defaults :/

Reply:
Well if we're changing the args from protocol to impl it won't be duping anyway and pdoc3 is inheriting the docs here

Reply:
Removed the ones you listed, if you think any more should be removed then send a list

Reply:
thats it